server/src/main/java/org/elasticsearch/action/support/FanOutListener.java

```diff
@@ -0,0 +1,305 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.action.support;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.util.concurrent.EsExecutors;
+import org.elasticsearch.common.util.concurrent.ListenableFuture;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.core.Nullable;
+
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * An {@link ActionListener} to which other {@link ActionListener} instances can subscribe, such that when this listener is completed it
+ * fans-out its result to the subscribed listeners.
+ *
+ * Similar to {@link ListenableActionFuture} and {@link ListenableFuture}, except that:
+ * <ul>
+ * <li>This implementation is not a {@link Future} so it has no support for a blocking wait for completion (although you can, for instance,
+ * subscribe a {@link PlainActionFuture} and block on that if desired).
+ * <li>The methods on this class are immune to interrupts.
+ * </ul>
+ */
+public final class FanOutListener<T> implements ActionListener<T> {
```

Comment:
This is something I've had in a development branch for a while. In due course we can use it to replace almost all usages of `ListenableActionFuture` and `ListenableFuture` because almost nowhere do we use these things as a future (i.e. in a blocking fashion). But it's useful here because of its uninterruptibility.